localstack/services/generic_proxy.py

```diff
@@ -592,136 +594,88 @@ def get_flask_ssl_context(cls, serial_number=None):
-        else:
-            forward_host = host
-        forward_path_root = self.forward_url.path
-        forward_path = path[len(self.base_url.path) :]  # strip base path
-
-        # avoid double slashes
-        if forward_path and not forward_path_root.endswith("/"):
-            if not forward_path.startswith("/"):
-                forward_path = "/" + forward_path
-
-        forward_url = forward_host + forward_path_root + forward_path
+    def register(self, forward_url):
+        """
+        Registers two routes at the edge router to implement either a path or a domain based routing.
+        We need two rules to match 1) the root path + no additional path 2) the root path + any other path.
+        This function is NOT called automatically, it is usually called in the leaf class' start up routine.
 
-        return urlparse(forward_url)
+        :param forward_url: The url which the proxy forwards to (from the initially passed base_url)
+        """
 
+        #   We MUST NOT create a catch all traffic rule
+        assert not (
+            (self._url.path == "" or self._url.path is None or self._url.path == "/")
+            and self._url.hostname == config.LOCALSTACK_HOSTNAME
+        )
+        from localstack.services.edge import ROUTER
 
-class EndpointProxy(ProxyListener):
-    def __init__(self, base_url: str, forward_url: str) -> None:
-        super().__init__()
-        self.forwarder = UrlMatchingForwarder(
-            base_url=base_url,
-            forward_url=forward_url,
+        self.rules.append(
+            ROUTER.add(
+                self._url.path or "/",
+                ProxyHandler(forward_url),
+                f"{self._url.hostname}<regex('(:.*)?'):port>",
+            )
+        )
+        self.rules.append(
+            ROUTER.add(
+                f"{self._url.path or ''}/<path:path>",
+                ProxyHandler(forward_url),
+                f"{self._url.hostname}<regex('(:.*)?'):port>",
+            )
         )
 
-    def forward_request(self, method, path, data, headers):
-        return self.forwarder.forward_request(method, path, data, headers)
+    def unregister(self):
+        for rule in self.rules:
+            from localstack.services.edge import ROUTER
 
-    def register(self):
-        ProxyListener.DEFAULT_LISTENERS.append(self)
+            ROUTER.remove_rule(rule)
+            LOG.debug("Removing router rule %s for %s", rule.rule, rule.host)
+        self.rules = []
 
-    def unregister(self):
+    def do_shutdown(self):
         try:
-            ProxyListener.DEFAULT_LISTENERS.remove(self)
-        except ValueError:
-            pass
+            self.unregister()
+        except KeyError as e:
+            # KeyError occurs when a rule is no longer present that is attempted to be removed
+            LOG.debug(e)
 
 
-class FakeEndpointProxyServer(Server):
+class FakeEndpointProxyServer(RegisteredProxyServer):
```

Comment:
What's this class for? It seems this abstraction (connecting the server and the router / proxy) is already handled in the `RegisteredProxyServer` class?